### PR TITLE
Template based challenges

### PR DESCRIPTION
**Pull request by KhalidRmb**
Includes the backend and frontend for the Template based challenge creation feature.

Through this new addition, challenge hosts can use pre configured challenges and get started with minimal setup.

@RishabhJain2018 @Ram81 @deshraj

### CODE REVIEW

File: frontend_v2/src/app/components/template-challenge-create/template-challenge-create.component.ts [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+	 	'end_date': null,
+	 	'description': null,
+	 	'evaluation_script': null,
+	 	'challenge_phases': null
+	 }
+
+	 challenge_phases = []
+
+	 challengePhaseData = {
+	 	'name': null,
```
Comment (Ram81) [inline frontend_v2/src/app/components/template-challenge-create/template-challenge-create.component.ts:39]: fix indentation

File: frontend_v2/src/app/components/template-challenge-create/template-challenge-create.component.ts [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+	 	'start_date': null,
+	 	'end_date': null
+	 }
+
+	 hostedChallengesRoute = '/challenges/me';
+	 hostTeamsRoute = '/teams/hosts';
+
+	 currentDateTime: Date;
+
+	constructor(
```
Comment (Ram81) [inline frontend_v2/src/app/components/template-challenge-create/template-challenge-create.component.ts:49]: fix indentation

File: frontend_v2/src/app/components/template-challenge-create/template-challenge-create.component.ts [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+    private router: Router,
+    private route: ActivatedRoute,
+    private challengeService: ChallengeService,
+    @Inject(DOCUMENT) private document,
+    private globalService: GlobalService,
+    private apiService: ApiService
+  ) {}
+
+	ngOnInit() {
+		if (this.authService.isLoggedIn()) {
```
Comment (Ram81) [inline frontend_v2/src/app/components/template-challenge-create/template-challenge-create.component.ts:60]: fix indentation

File: frontend_v2/src/app/components/template-challenge-create/template-challenge-create.component.ts [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+	      this.challengeService.currentHostTeam.subscribe((hostTeam) => {
+	      	this.hostTeam = hostTeam;
+		    if (!this.hostTeam) {
+		    	this.router.navigate([this.hostTeamsRoute], { queryParams: { template: true, templateId: this.id, templatePhases: this.phases } });
+			}
+		  });
+	    });
+
+	    for(var i = 0; i<this.phases; i++){
+	    	this.challengePhaseData['id'] = i+1;
```
Comment (Ram81) [inline frontend_v2/src/app/components/template-challenge-create/template-challenge-create.component.ts:82]: change to `this.challengePhaseData['id'] = i + 1;`

File: frontend_v2/src/app/components/template-challenge-create/template-challenge-create.component.ts [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+	}
+
+
+	createTemplateChallenge(){
+		if (this.challengeData != null) {
+			const FORM_DATA: FormData = new FormData();
+			this.challengeData['is_challenge_template'] = true;
+			this.challengeData['template_id'] = this.id
+			this.challengeData.challenge_phases = JSON.stringify(this.challenge_phases);
+			//FORM_DATA.append('data', JSON.stringify(this.challengeData));
```
Comment (Ram81) [inline frontend_v2/src/app/components/template-challenge-create/template-challenge-create.component.ts:100]: remove comment if not needed

File: frontend_v2/src/app/components/template-challenge-create/template-challenge-create.component.ts [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+			for(let key in this.challengeData){
+				FORM_DATA.append(key, this.challengeData[key]);
```
Comment (Ram81) [inline frontend_v2/src/app/components/template-challenge-create/template-challenge-create.component.ts:102]: fix indentation

File: frontend_v2/src/app/components/template-challenge-create/template-challenge-create.component.ts [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+		          this.globalService.showToast('error', "Sorry, something went wrong when creating the challenge. Please try again later: " + JSON.stringify(err.error) );
+		        },
+		        () => {}
+		      );
+		}
+		else {
+			this.globalService.showToast('error', 'Please fill all the given challenge details');
+		}
+	}
+}
```
Comment (Ram81) [inline frontend_v2/src/app/components/template-challenge-create/template-challenge-create.component.ts:123]: add newline

**Review by Ram81 — CHANGES_REQUESTED** [review]

File: settings/staging.py [hunk 8]
```diff
@@ -2,6 +2,8 @@
 
 ALLOWED_HOSTS = ["staging-evalai.cloudcv.org"]
 
+API_HOST_URL = "https://staging-evalai.cloudcv.org"
```
Comment (RishabhJain2018) [inline settings/staging.py:5]: Please remove it from here.

File: settings/prod.py [hunk 9]
```diff
@@ -11,6 +11,8 @@
     "evalapi.cloudcv.org",
 ]
 
+API_HOST_URL = "https://evalapi.cloudcv.org"
```
Comment (RishabhJain2018) [inline settings/prod.py:14]: Please remove it from here.

File: settings/common.py [hunk 10]
```diff
@@ -33,6 +33,7 @@
 
 ALLOWED_HOSTS = []
 
+API_HOST_URL = ""
```
Comment (RishabhJain2018) [inline settings/common.py:36]: Can we please get it from the environment and if the environment isn't having then use the default value `http://localhost:8000`

File: settings/dev.py [hunk 11]
```diff
@@ -9,6 +9,8 @@
 
 ALLOWED_HOSTS = ["*"]
 
+API_HOST_URL = "http://localhost:8000"
```
Comment (RishabhJain2018) [inline settings/dev.py:12]: Please remove it from here.

File: settings/test.py [hunk 12]
```diff
@@ -2,6 +2,8 @@
 
 DEBUG = False
 
+API_HOST_URL = "http://localhost:8000"
```
Comment (RishabhJain2018) [inline settings/test.py:5]: Please remove it from here.

**Review by RishabhJain2018 — CHANGES_REQUESTED** [review]

**Comment by Ram81** [comment]
@KhalidRmb can you fix the failing build

**Comment by KhalidRmb** [comment]
> @KhalidRmb can you fix the failing build
@Ram81 @RishabhJain2018 
Build has passed here: https://travis-ci.org/github/Cloud-CV/EvalAI/builds/725987642
Travis is not showing the new build here

**Comment by RishabhJain2018** [comment]
I restarted the build.

**Comment by Ram81** [comment]
@KhalidRmb on local I don't see any templates listed while creation. Is that not checked in?
Screen Shot 2020-09-10 at 9 54 20 PM

**Comment by Ram81** [comment]
@KhalidRmb for now please add 2 challenge templates in `seed.py` for dev work.

**Review by RishabhJain2018 — APPROVED** [review]

**Comment by RishabhJain2018** [comment]
@KhalidRmb Please let me know when this is ready after @Ram81's comment.

**Comment by KhalidRmb** [comment]
@Ram81 @RishabhJain2018 
Screenshot 2020-09-10 at 11 33 38 PM

**Review by RishabhJain2018 — APPROVED** [review]

**Pull request merged**